### PR TITLE
[RFC] Vudu 2.0, smaller pkg size and simpler api

### PR DESCRIPTION
**Pull request by dhunninghake**
January 20, 2019

# Summary

In this RFC, I’m proposing a modified approach to writing styles in JS with Vudu.

See an associated PR 👉🏼 https://github.com/dhunninghake/vudu/pull/23

Vudu 2.0...

- Is a smaller package size — `7.2kb` vs the current `18.6kb`.
- Takes a less opinionated approach to writing and composing JS objects.
- Exposes a simpler API for creating and interacting with styles.

# Basic example

The main difference is that objects passed to vudu do not need to be named. Like the previous version, it will generate a random class name and return it after appending the styles:

```js
import vudu from 'vudu';

export const Example = () => <div className={className} />;

const className = vudu({
  color: 'red',
  ':hover': {
    color: 'green',
  },
});
```

It is also possible to achieve a deterministic class name by passing a string as the first argument and currying your style object:

```js
import vudu from 'vudu';

export const Example = () => <div className="hi" />;

const className = vudu('.hi')({
  color: 'red',
  ':hover': {
    color: 'green',
  },
});
```

# Motivation

A few observations about the package after using it for some time:

- **Naming each style object is cumbersome** - previously it was required that any object passed to vudu needed a key. Over time this became a nuisance, especially when I just wanted styles generated without having to name them.

```js
const className = vudu({
  container: {
    // <---- this
    color: 'red',
  },
});
```

- **Composes keyword is unnecessary with es6** - one of the key concepts with vudu 1.0 was `@composes`. This was an array of pre-generated objects that could be imported and used in components. Instead of parsing this array, I've been using the spread operator to compose smaller atomic style objects.
- **The package src is hard to follow** - I was the one who wrote the initial version of vudu, and yet every time I come back to make updates or optimizations, I think "how the f does this work". So I started a rewrite one day and found that I could get nearly the same functionality with _way_ less code. Now, nearly all logic takes place in a single recursive function.

# Detailed design

### Breaking changes

The following features have been removed in Vudu 2.0, and the reasons why:

- **`v.composes`** - Support for this monolithic object of atomic styles is being removed to keep the package lean. This functionality can easily be achieved by writing a thin wrapper around vudu.
- **`@composes`** - As mentioned before, this feature can be achieved with simply spreading cached atomic style objects into an object passed to vudu.
- **`@font-face`** - Setting up typefaces is mostly a static operation. Do it once at the top of the app and then leave it alone. Because of this, a static stylesheet for fonts seems like a better option.
- **`@keyframes`** - Generating keyframes with JS can be very dynamic and powerful, but I rarely ever used this feature. I might add support for this feature back in, but still on the fence.
- **Custom style sheet** - Only included this feature for testing purposes. Vudu 2.0 takes a different approach to testing.

### Stayed the same

- **Random class names** - Vudu continues to generate and return a random class names for unnamed style objects passed to it.
- **Vendor prefixing** - The `inline-style-prefixer` package accounts for `5.8kb` of the `7.2kb` vudu is now becoming. It’s a lot, but I consider it a worthy cost in order to build robust components which stand up to varying browser css feature support.

### New additions

- **Deterministic class names** - This means it’s possible to assign a class name to style objects. Additionally, this opens the door to setting css for global elements.
- **New methods** - `.reset()` and `.css()` are now available to empty the current stylesheet or read from it.

# Migration Path

A look at how to move forward regarding the breaking changes above:

### `v.composes`

Currently the only way to get this monolithic object would be to write a wrapper around Vudu 2.0. It would be pretty simple. However, although this functionality is being removed from the base package, we could consider a secondary package in the repo that offers support for this object. That way the base stays lean while not removing this altogether. It might look something like:

```js
import composes from 'vudu/composes';
```

### `@composes`

This one is easy, it can be achieved with es6.

#### Before

```js
const className = v({
  container: {
    @composes: [red, pt2]
  }
});
```

#### After

```js
const className = vudu({
  ...red,
  ...pt2,
});
```

### `@font-face`

Create a static stylesheet with font declarations.

### `@keyframes`

Similar to font face, these can be added to a static sheet. It’s possible support for this will be added back in.

### `Custom style sheet`

The only reason this was supported in v1 was so that I could pass a custom style sheet in tests. I’ve now taken a different approach in testing to simply unit test the string output the style object which makes a custom stylesheet unnecessary.

# Looking ahead

One goal I want to keep in mind with vudu is that it continues to offer an option for inline styles with JS outside the context of a modern framework. Whether your build is in React, Vue, Ember, the next hot JS framework, or plain old JS, I want it to be flexible enough to adapt to your use case. Any refinements to the inner workings or the API in the future will be in the service of making it faster and more accessible to more developers.

<3

### CODE REVIEW

**Comment by dhunninghake** [comment]
@hhff @sepowitz ready for another 👀 whenever you two find some time! I'm curious about a couple things from your perspective:

- How difficult would it be to upgrade the apps you've got using vudu 1.x to 2.0? Is there anything that can be changed/added to make that easier?
- Are there use cases for vudu that it doesn't yet support? Seb mentioned a couple in the rewrite PR, but I'd love to know more and come up with a plan to support them.

**Comment by hhff** [comment]
Hi big D! We had a bunch of sites go down due to a CMS provider deprecating, so I've been timepoor lately.

Will review ASAP!

File: rfcs/001-rewrite.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+```
+
+#### After
+
+```js
+const className = vudu({
+  ...red,
+  ...pt2,
+});
+```
```
Comment (hhff) [inline rfcs/001-rewrite.md]: Hi D! Just got a chance to review this.

The main thing that's going to make this hard for us to migrate over is that 

```
const styles = v({
  container: {},
  innerWrapper: {}
})
```

is referred to in all our DOM like

```
<div className={styles.container}>
  <div className={styles.innerWrapper}></div>
</div>
```

Using the new API, do you have any ideas how we'd be able to migrate our codebases (some of these have upwards of 100 - 200+ vudu calls) without having to rewrite our entire DOM?

File: rfcs/001-rewrite.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+A look at how to move forward regarding the breaking changes above:
+
+### `v.composes`
+
+Currently the only way to get this monolithic object would be to write a wrapper around Vudu 2.0. It would be pretty simple. However, although this functionality is being removed from the base package, we could consider a secondary package in the repo that offers support for this object. That way the base stays lean while not removing this altogether. It might look something like:
+
+```js
+import composes from 'vudu/composes';
+```
```
Comment (hhff) [inline rfcs/001-rewrite.md]: Cool

File: rfcs/001-rewrite.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+- **Naming each style object is cumbersome** - previously it was required that any object passed to vudu needed a key. Over time this became a nuisance, especially when I just wanted styles generated without having to name them.
+
+```js
+const className = vudu({
+  container: {               // <---- this
+    color: 'red',
+  },
+});
+```
```
Comment (hhff) [inline rfcs/001-rewrite.md]: I tend to disagree that naming each style object is cumbersome - you have to name objects at some point, be it here, or in the destructured array (`const [container, innerWrapper] = vudu`). 

---

My main concern with this API is that if you were to insert an object into the call, like this:
```
--- const [c1, c2] = vudu({ color: 'red' }, { color: 'blue' });
+++ const [c1, c2, c3] = vudu({ color: 'red' }, {color: 'green'}, { color: 'blue' });
```

You've accidentally changed the names of all of your styles. This makes arranging your styles inside of the vudu object really arduous as your app evolves.

As a side, this is why array-type Enums are generally considered problematic when compared to indexed Enums in database design:

```
enum status: [:shipped, :delivered]
```
^^is dangerous, where as

```
enum status: {
  shipped: 0
  delivered: 1
}
```
Is safer. Adding a value forces you to index it in the second example, making it far less likely you'll ruin your database by adding an entry.

Thats something I liked about the Vudu API! it was easy to reorganize your styles within the object without breaking your DOM

File: rfcs/001-rewrite.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  }
+});
+```
+
+#### After
+
+```js
+const className = vudu({
+  ...red,
+  ...pt2,
```
Comment (hhff) [inline rfcs/001-rewrite.md]: Feels a tad more verbose than the original inline format... but I do still prefer the new syntax

File: rfcs/001-rewrite.md (hunk 1 above)
Comment (dhunninghake) [reply rfcs/001-rewrite.md]: Hugh! I'm sorry this has taken so long to respond 😅. I'm just now getting back into the vudu headspace. I think the simplest way forward for the example above would be to move the vudu calls down a level:

```js
const styles = {
  container: vudu({}),
  innerWrapper: vudu({})
})
```

That way you don't need to change anything about the DOM.

Sidenote: one of the motivations for changing the API to _not_ handle custom names for objects was because it made parsing the style objects much harder. This way, vudu will always only wrap styles that need to be parsed and added to a stylesheet. I think this gives more flexibility to the app authors to determine if/how/when vudu objects are assigned to variables.

File: rfcs/001-rewrite.md (hunk 3 above)
Comment (dhunninghake) [reply rfcs/001-rewrite.md]: You're right! It doesn't seem safe to assume the index won't change (i.e. append only). With fresh eyes I'm wondering now if vudu handling multiple objects at once is even worth it given the probability of it causing bugs.

Wdyt about me removing this altogether and using a strategy like the one I mentioned above, where the author can set up their own object of classnames?

**Comment by dhunninghake** [comment]
@hhff @sepowitz Question for you guys: do you use snapshot tests with your components? I personally don't love snaps for many reasons, but in a case like upgrading apps from vudu v1 to v2 you could throw some snapshot tests in there and make sure nothing looks too crazy before merging the upgrade. Would love to know if there's anything blocking on your side (code concerns or otherwise) for me merging this new version!
